src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java

```diff
@@ -249,27 +253,57 @@ private CompletableFuture<Void> registerFuture() {
   }
 
   /**
-   * Kick off a flush job and distribute the tasks if one of the following conditions is met:
-   * <li>Flush is forced by the users
-   * <li>One or more buffers have reached the flush size
-   * <li>Periodical background flush when a time interval has reached
+   * Kick off a flush job and distribute the tasks. The flush service behaves differently based on
+   * the max chunks in blob:
+   *
+   * <ul>
+   *   <li>The max chunks in blob is not 1 (interleaving is allowed), every channel will be flushed
+   *       together if one of the following conditions is met:
+   *       <ul>
+   *         <li>Flush is forced by the users
+   *         <li>One or more buffers have reached the flush size
+   *         <li>Periodical background flush when a time interval has reached
+   *       </ul>
+   *   <li>The max chunks in blob is 1 (interleaving is not allowed), a channel will be flushed if
+   *       one of the following conditions is met:
+   *       <ul>
+   *         <li>Flush is forced by the users
+   *         <li>One or more buffers with the same target table as the channel have reached the
+   *             flush size
+   *         <li>Periodical background flush of the target table when a time interval has reached
+   *       </ul>
+   * </ul>
    *
    * @param isForce
    * @return Completable future that will return when the blobs are registered successfully, or null
    *     if none of the conditions is met above
    */
   CompletableFuture<Void> flush(boolean isForce) {
-    long timeDiffMillis = System.currentTimeMillis() - this.lastFlushTime;
+    long currentTime = System.currentTimeMillis();
+    long timeDiffMillis = currentTime - this.lastFlushTime;
+
+    Set<String> tablesToFlush =
+        this.channelCache.entrySet().stream()
+            .filter(
+                entry ->
+                    isForce
+                        || currentTime - this.channelCache.getLastFlushTime(entry.getKey())
+                            >= this.owningClient.getParameterProvider().getCachedMaxClientLagInMs()
+                        || this.channelCache.getNeedFlush(entry.getKey()))
+            .map(Map.Entry::getKey)
+            .collect(Collectors.toSet());
+
+    // Flush every table together when it's interleaving chunk is allowed
+    if (this.owningClient.getParameterProvider().getMaxChunksInBlobAndRegistrationRequest() != 1
+        && !tablesToFlush.isEmpty()) {
+      tablesToFlush.addAll(
+          this.channelCache.entrySet().stream().map(Map.Entry::getKey).collect(Collectors.toSet()));
+    }
```

Comment:
Why do we need to do this, If the previous code block already picked up the minimal set of tables needing flush?

Reply:
+1, even if interleaving is enabled, I'd prefer to keep the above logic for flushing and wait until the MaxClientLag for each channel

Reply:
I aimed to maintain the original interleaving behavior, where all channels are flushed if any channel needs it. With independent flushing intervals, we might miss the chance to combine multiple chunks into the same BDEC. A potential workaround is to discretize timestamps and reduce jitter on `lastFlushTime` in interleaving mode. This can increase the chances of combining multiple chunks into the same blob. What do you think?